python/src/iceberg/expressions/base.py

```diff
@@ -137,6 +74,28 @@ def __invert__(self) -> "BooleanExpression":
         ...
 
 
+class BoundPredicate(ABC):
+    def __init__(self, left, right):
+        """A concrete predicate must have a `left` and `right`"""
+        self._left = left
+        self._right = right
```

Comment:
The usage of `left` and `right` here is somewhat confusing for me. Is this something that I missed that differs (or is renamed) from the Java library?

Or possibly `left` and `right` correspond tot the java `term` and `literal` arguments of `BoundLiteralPredicate`?

Reply:
> Or possibly left and right correspond tot the java term and literal arguments of BoundLiteralPredicate?

Yes exactly. The plan is to not have a generic `BoundLiteralPredicate` that uses an `Operation` enum + a switch statement. Instead we'll have a different class for every operation, so `In(BooleanExpression, BoundPredicate)`, `LessThan(BooleanExpression, BoundPredicate)`, `GreaterThan(BooleanExpression, BoundPredicate)`, etc.

Since we're working with binary expression trees, I think it makes sense to use `left` and `right` here to keep the visitor logic intuitive, even though this deviates from the java `term` and `literal` argument names. IMO the fact that `left` is always a `term` and `right` is always a `literal` can simply be described through typehints and the docstring.

Reply:
Oh ok I see that left is the reference and right is the literal(s).

Could we possibhly name this something more specific than `left` / `right`?

Reply:
Does `ref` and `literal` sound better? That's what it originally was in this PR. The visit function would then look like:
```py
@visit.register(UnboundPredicate)
def _(obj: UnboundPredicate, visitor: BooleanExpressionVisitor[T]) -> T:
    """Visit an unbound predicate with a concrete BooleanExpressionVisitor"""
    ref_result: T = visit(obj.ref, visitor=visitor)
    literal_result: T = visit(obj.literal, visitor=visitor)
    return visitor.visit_unbound_predicate(ref_result=ref_result, literal_result=literal_result)
```

Reply:
Let's not worry about the visitor just yet. I think we would probably pass the actual predicate rather than deconstructing it. Otherwise we'd need to reintroduce operation to pass it in.

I agree that this should be more specific about its children. I think it should use `term` (always left) and `literal`.